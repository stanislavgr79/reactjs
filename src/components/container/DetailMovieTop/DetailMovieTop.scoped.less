.movie_detail {
    background-color: #000;
    .model_content {
        position: relative;
        padding-bottom: 20px;
        margin-bottom: 20px;
        height: 100%;
        pointer-events: auto;
        .section_icon-close {
            position: absolute;
            line-height: 1rem;
            top: 8px;
            right: 8px;
            margin-right: 20px;
            margin-top: 5px;
            .btn_close.icon_small {
                padding: 0;
                background-size: cover;
                line-height: 1rem;
                width: 1rem;
                height: 1rem;
            }
        }
    }
}

.detail_movie-card {
    display: flex;
    flex-direction: row;
    position: relative;
    width: auto;
    height: 88%;
    padding-right: 20px;
    margin: 20px 0;
    max-height: 290px;
    .section_image {
        min-width: 200px;
        align-self: center;
        position: relative;
        img {
            max-height: 100%;
            max-width: -webkit-fill-available;
        }
    }
    .section-detail {
        align-self: center;
        padding-left: 40px;
        color: red;
        .title_reiting {
            display: flex;
            flex-direction: row;
            font-size: 38px;
            .movie-title {
                color: #fff;
                padding-right: 10px;
            }
            .vote_average {
                border-radius: 50%;
                border-color: #bbb;
                box-sizing: 2px;
                color: green;
                font-size: 24px;
                border: 0.5px solid #bbbbbb;
                text-align: center;
                padding: 6px 5px 0 5px;
                width: 2.1em;
                height: 2.1em;
                border-radius: 50%;
                margin-top: 8px;
            }
        }
        .tagline {
            color: #bbb;
            font-size: 16px;
            padding: 5px 0;
        }
        .release_runtime {
            display: flex;
            flex-direction: row;
            color: #f65261;
            padding: 10px 0;
            font-size: 22px;
            .release {
                padding-right: 20px;
            }
        }
        .movie_overview {
            color: #BBB;
            font-size: 16px;
        }
    }
}