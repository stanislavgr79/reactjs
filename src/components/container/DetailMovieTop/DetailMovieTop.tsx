import React, { useCallback } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { AppState } from '../../../redux/redux-store';
import { updateSelectedMovieId } from '@redux/actions/movies-actions';

import Button from '@components/atom/Button';
import Sitename from '@components/atom/SiteName/SiteName';

import './DetailMovieTop.scoped.less';

export default function DetailMoviePopup(): JSX.Element {
  const dispatch = useDispatch();

  const store = useSelector((store: AppState) => {
    return {
      movie: store.moviesStore.movie,
    };
  });

  const resetActiveMovieId = useCallback(() => {
    dispatch(updateSelectedMovieId(0));
  }, [dispatch]);

  const year = store.movie?.release_date.split('-')[0];

  return (
    <section className="movie_detail">
      <div className="model_content container-md">
        <Sitename />
        <div className="section_icon-close">
          <Button
            buttonType="button"
            className="btn_close icon_small"
            onClick={resetActiveMovieId}
          />
        </div>
        <div className="detail_movie-card">
          <div className="section_image">
            <img
              src={store.movie?.poster_path}
              className="img-fluid"
              alt={store.movie?.title}
              onClick={(e) => e.preventDefault()}
            />
          </div>
          <div className="section-detail">
            <div className="title_reiting">
              <div className="movie-title">{store.movie?.title}</div>
              <div className="vote_average">{store.movie?.vote_average}</div>
            </div>
            <div className="tagline">{store.movie?.tagline}</div>
            <div className="release_runtime">
              <div className="release">{year}</div>
              <div className="runtime">{store.movie?.runtime} min</div>
            </div>
            <div className="movie_overview">{store.movie?.overview}</div>
          </div>
        </div>
      </div>
    </section>
  );
}
