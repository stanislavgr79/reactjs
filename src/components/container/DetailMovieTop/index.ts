import DetailMovieTop from './DetailMovieTop';

export default DetailMovieTop;
