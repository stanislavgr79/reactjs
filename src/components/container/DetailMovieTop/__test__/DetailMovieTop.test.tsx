import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { createMemoryHistory } from 'history';
import ReactRouter from 'react-router';

import DetailMovieTop from '../DetailMovieTop';

import { render, screen, fireEvent } from '@testing-library/react';
import {
  MoviesActionTypes,
  UPDATE_SELECTED_MOVIE_ID,
} from '../../../../redux/types/movies-reducer-types';

jest.mock('../../../atom/SiteName', () => {
  return {
    __esModule: true,
    default: () => {
      return <div className="siteName"></div>;
    },
  };
});
jest.mock('../../../../redux/actions/movies-actions', () => ({
  updateSelectedMovieId: (): MoviesActionTypes => {
    return {
      type: UPDATE_SELECTED_MOVIE_ID,
      payload: 0,
    };
  },
}));

describe('DetailMovieTop', () => {
  const movie = {
    id: 12345,
    title: 'film title',
    release_date: '2020-01-02',
    poster_path: null,
    genres: ['crime'],
    overview: 'overview',
    runtime: 66,
    vote_average: 5.5,
    tagline: 0,
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('resetActiveMovieId function with empty query in search value', () => {
    const initialState = {
      moviesStore: { movie: movie },
      searchStore: { search: { value: '' } },
    };
    const mockStore = configureStore();
    const store = mockStore(initialState);

    const mockLocation = {
      pathname: '/movies',
      search: '',
      state: null,
      hash: '',
    };
    const history = createMemoryHistory();
    history.location = mockLocation;
    jest.spyOn(ReactRouter, 'useLocation').mockReturnValue(mockLocation);
    const replaceHistorySpy = jest.spyOn(ReactRouter, 'useHistory').mockReturnValue(history);

    const tree = render(
      <Provider store={store}>
        <DetailMovieTop />
      </Provider>,
    );

    const expectedActions = [{ type: 'UPDATE_SELECTED_MOVIE_ID', payload: 0 }];

    expect(tree.baseElement).toMatchSnapshot();

    fireEvent.click(screen.getByRole('button', { name: '' }));

    expect(store.getActions()).toEqual(expectedActions);
    expect(replaceHistorySpy).toHaveBeenCalled();
  });

  test('resetActiveMovieId function with query in search value', () => {
    const initialState = {
      moviesStore: { movie: movie },
      searchStore: { search: { value: 'qwerty' } },
    };
    const mockStore = configureStore();
    const store = mockStore(initialState);

    const mockLocation = {
      pathname: '/movies',
      search: 'qwerty',
      state: null,
      hash: '',
    };
    const history = createMemoryHistory();
    history.location = mockLocation;
    jest.spyOn(ReactRouter, 'useLocation').mockReturnValue(mockLocation);
    const replaceHistorySpy = jest.spyOn(ReactRouter, 'useHistory').mockReturnValue(history);

    const tree = render(
      <Provider store={store}>
        <DetailMovieTop />
      </Provider>,
    );

    const expectedActions = [{ type: 'UPDATE_SELECTED_MOVIE_ID', payload: 0 }];

    expect(tree.baseElement).toMatchSnapshot();

    fireEvent.click(screen.getByRole('button', { name: '' }));

    expect(store.getActions()).toEqual(expectedActions);
    expect(replaceHistorySpy).toHaveBeenCalled();
  });
});
