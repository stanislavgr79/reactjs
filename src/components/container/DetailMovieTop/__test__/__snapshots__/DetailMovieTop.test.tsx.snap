// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailMovieTop resetActiveMovieId function with empty query in search value 1`] = `
<body>
  <div>
    <section
      class="movie_detail"
    >
      <div
        class="model_content container-md"
      >
        <div
          class="siteName"
        />
        <div
          class="section_icon-close"
        >
          <button
            class="btn button btn_close icon_small"
            id=""
            type="button"
          />
        </div>
        <div
          class="detail_movie-card"
        >
          <div
            class="section_image"
          >
            <img
              alt="film title"
              class="img-fluid"
              src="[object Object]"
            />
          </div>
          <div
            class="section-detail"
          >
            <div
              class="title_reiting"
            >
              <div
                class="movie-title"
              >
                film title
              </div>
              <div
                class="vote_average"
              >
                5.5
              </div>
            </div>
            <div
              class="tagline"
            >
              0
            </div>
            <div
              class="release_runtime"
            >
              <div
                class="release"
              >
                2020
              </div>
              <div
                class="runtime"
              >
                66
                 min
              </div>
            </div>
            <div
              class="movie_overview"
            >
              overview
            </div>
          </div>
        </div>
      </div>
    </section>
  </div>
</body>
`;

exports[`DetailMovieTop resetActiveMovieId function with query in search value 1`] = `
<body>
  <div>
    <section
      class="movie_detail"
    >
      <div
        class="model_content container-md"
      >
        <div
          class="siteName"
        />
        <div
          class="section_icon-close"
        >
          <button
            class="btn button btn_close icon_small"
            id=""
            type="button"
          />
        </div>
        <div
          class="detail_movie-card"
        >
          <div
            class="section_image"
          >
            <img
              alt="film title"
              class="img-fluid"
              src="[object Object]"
            />
          </div>
          <div
            class="section-detail"
          >
            <div
              class="title_reiting"
            >
              <div
                class="movie-title"
              >
                film title
              </div>
              <div
                class="vote_average"
              >
                5.5
              </div>
            </div>
            <div
              class="tagline"
            >
              0
            </div>
            <div
              class="release_runtime"
            >
              <div
                class="release"
              >
                2020
              </div>
              <div
                class="runtime"
              >
                66
                 min
              </div>
            </div>
            <div
              class="movie_overview"
            >
              overview
            </div>
          </div>
        </div>
      </div>
    </section>
  </div>
</body>
`;
